### PR TITLE
Add deprecation logging and support for binary compatibility to upgraded methods

### PR DESCRIPTION
**Pull request by asodja**
We now add deprecation logging by default to upgraded properties, by default we generate something like:
```
DeprecationLogger.deprecateProperty(Task.class, "maxErrors")
    .withContext("Property was automatically upgraded to the lazy version.")
    .startingWithGradle9("Property is replaced with lazy version.")
    .undocumented()
    .nagUser();
``` 

It can be modified with `@UpgradedDeprecation` annotation, e.g.:
```
@UpgradedProperty(deprecation = @UpgradedDeprecation(
  enabled = true,
  removedIn = 9,
  withUpgradeGuideMajorVersion = 7, 
  withUpgradeGuideSection = "some_section"
))
```

----
Additionaly this PR adds additional feature to `@UpgradedProperty`. 
You can specify binary compatibility as:
```
@UpgradedProperty(binaryCompatibility = METHODS_KEPT)
```

This can be useful since `@ReplacedBy` methods are not removed, so our binary compatibility is aware of that.

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

File: platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/annotations/UpgradedProperty.java [hunk 1]
```diff
@@ -43,6 +45,25 @@
 
     UpgradedAccessor[] originalAccessors() default {};
 
+    BinaryCompatibility binaryCompatibility() default METHODS_REMOVED;
```
Comment (lptr) [inline platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/annotations/UpgradedProperty.java:48]: ❓ WDYT about calling these "accessors" instead of "methods"? We already use `originalAccessors` and `UpgradedAccessor`.

File: platforms/core-runtime/internal-instrumentation-processor/build.gradle.kts [hunk 2]
```diff
@@ -42,6 +42,7 @@ dependencies {
     implementation(libs.jacksonDatabind)
 
     implementation(project(":base-services"))
+    implementation(project(":logging"))
```
Comment (lptr) [inline platforms/core-runtime/internal-instrumentation-processor/build.gradle.kts:45]: ❓ Why is this necessary? Is this because of the deprecation logger?

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java [hunk 3]
```diff
@@ -107,57 +109,100 @@ private static List<AnnotationSpec> getAnnotations(PropertyUpgradeRequestExtra i
...
+            depreactionBuilder.add(".replaceWith($S)\n", newPropertyName);
+        }
+
+        if (deprecationSpec.getRemovedIn() == -1) {
+            depreactionBuilder.add(".startingWithGradle9($S)\n", "Property is replaced with lazy version.");
+        } else if (deprecationSpec.getRemovedIn() == 9) {
+            depreactionBuilder.add(".willBeRemovedInGradle9()\n");
+        } else if (deprecationSpec.getRemovedIn() != -1) {
+            throw new UnsupportedOperationException("Only unset or 9 is currently supported for removedIn, but was: " + deprecationSpec.getRemovedIn());
+        }
```
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java:151]: Makes sence, used enum

File: platforms/core-runtime/internal-instrumentation-processor/build.gradle.kts (hunk 2 above)
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/build.gradle.kts]: Removed, we reference `DeprecationLogger`, but we actually don't need to reference the type, we can just use a string.

File: platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/annotations/UpgradedProperty.java (hunk 1 above)
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/annotations/UpgradedProperty.java]: Makes sense, renamed this and bunch of things from "methods" to "accessors".

**Comment by asodja** [comment]
@bot-gradle test this

File: platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/annotations/UpgradedDeprecation.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    /**
+     * Corresponds to .withDslReference() in the DeprecationLogger
+     *
+     * WithDslReference has lower priority than withUpgradeGuideSection
+     */
+    boolean withDslReference() default false;
+
+    enum RemovedIn {
+        GRADLE9,
+        UNSET
```
Comment (lptr) [inline platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/annotations/UpgradedDeprecation.java:62]: ❓ Should this be called something else, like `NEVER` or `UNSPECIFIED`?

**Review by lptr — APPROVED** [review]
LGTM.

File: platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/annotations/UpgradedDeprecation.java (hunk 4 above)
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/annotations/UpgradedDeprecation.java:62]: Makes sense, I like `UNSPECIFIED`.

**Comment by asodja** [comment]
@bot-gradle test QF

**Comment by asodja** [comment]
@bot-gradle clean
